aiosmtpd/tests/conftest.py

```diff
@@ -0,0 +1,124 @@
+import pytest
+import asyncio
+import inspect
+
+from aiosmtpd.controller import Controller
+from aiosmtpd.handlers import Sink
+from contextlib import suppress
+from smtplib import SMTP as SMTPClient
+from ..testing.helpers import DecodingController
+from typing import Any, Callable, Dict, NamedTuple, Optional, Type
+
+
+class HostPort(NamedTuple):
+    host: str = "localhost"
+    port: int = 8025
+
+
+class Global:
+    SrvAddr: HostPort = HostPort()
+
+    @classmethod
+    def set_addr_from(cls, contr: Controller):
+        cls.SrvAddr = HostPort(contr.hostname, contr.port)
+
+
+@pytest.fixture
+def get_controller(request) -> Callable[..., Controller]:
+    marker = request.node.get_closest_marker("controller_data")
+    if marker:
+        markerdata = marker.kwargs or {}
+    else:
+        markerdata = {}
+
+    def getter(
+        handler,
+        default: Optional[Type[Controller]] = Controller,
+        server_kwargs: Dict[str, Any] = None,
+    ) -> Controller:
+        assert not inspect.isclass(handler)
+        class_: Type[Controller] = markerdata.get("class_", default)
+        if class_ is None:
+            raise RuntimeError(
+                f"Fixture '{request.fixturename}' needs controller_data to specify "
+                f"what class to use"
+            )
+        ip_port: HostPort = markerdata.get("host_port", HostPort())
+        return class_(
+            handler,
+            hostname=ip_port.host,
+            port=ip_port.port,
+            server_kwargs=server_kwargs,
+        )
+
+    return getter
```

Comment:
This fixture allows per-test-case selection of Controller without having to write separate fixtures. Usually used if there's a certain Handler we want to test against several Controllers.